dpcpp/solver/idr_kernels.dp.cpp

```diff
@@ -55,11 +66,727 @@ namespace dpcpp {
 namespace idr {
 
 
+constexpr int default_block_size = 256;
+constexpr int default_dot_dim = 16;
+constexpr int default_dot_size = default_dot_dim * default_dot_dim;
+
+
+template <typename ValueType>
+void initialize_m_kernel(size_type subspace_dim, size_type nrhs,
+                         ValueType *__restrict__ m_values, size_type m_stride,
+                         stopping_status *__restrict__ stop_status,
+                         sycl::nd_item<3> item_ct1)
+{
+    const auto global_id = thread::get_thread_id_flat(item_ct1);
+    const auto row = global_id / m_stride;
+    const auto col = global_id % m_stride;
+
+    if (global_id < nrhs) {
+        stop_status[global_id].reset();
+    }
+
+    if (row < subspace_dim && col < nrhs * subspace_dim) {
+        m_values[row * m_stride + col] =
+            (row == col / nrhs) ? one<ValueType>() : zero<ValueType>();
+    }
+}
+
+template <typename ValueType>
+void initialize_m_kernel(dim3 grid, dim3 block, size_t dynamic_shared_memory,
+                         sycl::queue *stream, size_type subspace_dim,
+                         size_type nrhs, ValueType *m_values,
+                         size_type m_stride, stopping_status *stop_status)
+{
+    stream->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                initialize_m_kernel(subspace_dim, nrhs, m_values, m_stride,
+                                    stop_status, item_ct1);
+            });
+    });
+}
+
+
+template <size_type block_size, typename ValueType>
+void orthonormalize_subspace_vectors_kernel(
+    size_type num_rows, size_type num_cols, ValueType *__restrict__ values,
+    size_type stride, sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, block_size> &reduction_helper_array,
+    remove_complex<ValueType> *reduction_helper_real)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+
+    ValueType *__restrict__ reduction_helper = reduction_helper_array;
```

Comment:
@upsj is it what you means?
also the calling, 
```
cgh.parallel_for(
            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
                orthonormalize_subspace_vectors_kernel<block_size>(
                    num_rows, num_cols, values, stride, item_ct1,
                    *reduction_helper_array_acc_ct1.get_pointer(),
                    reduction_helper_real_acc_ct1.get_pointer().get());
            });
```
but I am a little unsure.
Is there any reason such that dpct do not use this way?

Reply:
If you pass only the complex reduction helper, you can use
```
ValueType *__restrict__ reduction_helper = reduction_helper_array;
remove_complex<ValueType>* real_reduction_helper = reinterpret_cast<remove_complex<ValueType>*>(reduction_helper);
```
That should be fine.

Reply:
I mean the reference part

Reply:
yes

Reply:
does real_reduction_helper not lead issue with reduction_helper?
reduction_helper thinks no one overlap on it, but real_reduction_helper does.

Reply:
That's why it's important to set the `__restrict__` on the first variable, but not the second! The compiler can't assume that pointers derived from the same pointer don't overlap.

Reply:
the second is overlapped on the first, so I thought it destroys the first `__restrict__` property?
Might the first be arranged after the second (if we do not consider other dependency) because the first is `__restrict__`?